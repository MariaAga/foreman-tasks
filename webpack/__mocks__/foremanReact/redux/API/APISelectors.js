export const selectAPIResponse = (state, key) => ({
  data: {
    text: 'some-data',
    key,
  },
});

export const selectAPIStatus = (state, key) => 'PENDING';
export const selectAPIError = (state, key) => ({ error: `${key} ERRROR` });
