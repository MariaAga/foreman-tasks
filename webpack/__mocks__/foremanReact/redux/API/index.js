export const API = {
  get: jest.fn(),
  put: jest.fn(),
  post: jest.fn(),
  delete: jest.fn(),
  patch: jest.fn(),
};

export const get = data => ({ type: 'some-type', ...data });
