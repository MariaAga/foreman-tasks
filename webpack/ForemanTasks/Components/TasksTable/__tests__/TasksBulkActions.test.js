import { testActionSnapshotWithFixtures } from '@theforeman/test';
import { API } from 'foremanReact/redux/API';
import {
  bulkCancelById,
  bulkCancelBySearch,
  bulkResumeById,
  bulkResumeBySearch,
  bulkForceCancelById,
  bulkForceCancelBySearch,
} from '../TasksBulkActions';

jest.mock('foremanReact/components/common/table', () => ({
  getTableItemsAction: jest.fn(controller => controller),
}));

jest.mock('foremanReact/redux/API');

const task = {
  id: 'some-id',
  name: 'some-name',
};

const fixtures = {
  'handles bulkResumeById requests that fail': () => {
    const selected = [{ ...task, isResumable: true }];

    API.post.mockImplementation(() =>
      Promise.reject(new Error('Network Error'))
    );
    return bulkResumeById({ selected, url: 'some-url' });
  },
  'handles resumable bulkResumeById requests': () => {
    const selected = [{ ...task, isResumable: true }];

    API.post.mockImplementation(() => ({
      data: {
        resumed: [{ action: 'I am resumed' }],
        failed: [{ action: 'I am failed' }],
      },
    }));
    return bulkResumeById({ selected, url: 'some-url' });
  },
  'handles bulkCancelById requests': () => {
    const selected = [{ ...task, isCancellable: true }];

    API.post.mockImplementation(() => ({
      data: {
        cancelled: [{ action: 'I am cancelled' }],
      },
    }));
    return bulkCancelById({ selected, url: 'some-url' });
  },
  'handles skipped bulkResumeById requests': () => {
    const selected = [{ ...task, isResumable: true }];

    API.post.mockImplementation(() => ({
      data: {
        skipped: [{ action: 'I am skipped' }],
      },
    }));
    return bulkResumeById({ selected, url: 'some-url' });
  },
  'handles skipped bulkCancelById requests': () => {
    const selected = [{ ...task, isCancellable: true }];

    API.post.mockImplementation(() => ({
      data: {
        skipped: [{ action: 'I am skipped' }],
      },
    }));
    return bulkCancelById({ selected, url: 'some-url' });
  },

  'handles bulkForceCancelById requests': () => {
    const selected = [{ ...task, isCancellable: true }];

    API.post.mockImplementation(() => ({
      data: {
        stopped_length: 2,
        skipped_length: 4,
      },
    }));
    return bulkForceCancelById({ selected, url: 'some-url' });
  },

  'handles bulkForceCancelById requests that fail': () => {
    const selected = [{ ...task, isResumable: true }];

    API.post.mockImplementation(() =>
      Promise.reject(new Error('Network Error'))
    );
    return bulkForceCancelById({ selected, url: 'some-url' });
  },

  'handles bulkCancelById requests that are not cancellable': () => {
    const selected = [{ ...task, isCancellable: false }];
    return bulkCancelById({ selected, url: 'some-url' });
  },
  'handles bulkResumeById requests that are not resumable': () => {
    const selected = [{ ...task, isResumable: false, isCancellable: false }];
    return bulkResumeById({ selected, url: 'some-url' });
  },

  'handles bulkForceCancelById requests that are stopped': () => {
    const selected = [{ ...task, isResumable: false, state: 'stopped' }];
    return bulkForceCancelById({ selected, url: 'some-url' });
  },

  'handles bulkCancelBySearch requests': () => {
    API.post.mockImplementation(() => ({
      data: {
        cancelled: [{ action: 'I am cancelled' }],
        skipped: [{ action: 'I am skipped' }],
      },
    }));
    return bulkCancelBySearch({
      query: { search: {} },
      url: 'some-url',
      parentTaskID: 'parent',
    });
  },

  'handles bulkResumeBySearch requests': () => {
    API.post.mockImplementation(() => ({
      data: {
        cancelled: [{ action: 'I am cancelled' }],
        skipped: [{ action: 'I am skipped' }],
        failed: [{ action: 'I am failed' }],
      },
    }));
    return bulkResumeBySearch({
      query: { search: {} },
      url: 'some-url',
      parentTaskID: 'parent',
    });
  },
  'handles bulkForceCancelBySearch requests': () =>
    bulkForceCancelBySearch({
      query: { search: {} },
      url: 'some-url',
      parentTaskID: 'parent',
    }),
};

describe('TasksTable bulk actions', () => {
  testActionSnapshotWithFixtures(fixtures);
});
