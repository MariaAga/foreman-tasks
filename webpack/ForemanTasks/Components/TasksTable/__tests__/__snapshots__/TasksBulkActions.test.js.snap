// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksTable bulk actions handles bulkCancelById requests 1`] = `
Array [
  Array [
    Object {
      "type": "TASKS_CANCEL_REQUEST",
    },
  ],
  Array [
    Object {
      "type": "TASKS_CANCEL_SUCCESS",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "message": Object {
          "message": "I am cancelled Task execution was cancelled",
          "type": "success",
        },
      },
      "type": "TOASTS_ADD",
    },
  ],
  Array [
    "TASKS_TABLE",
  ],
  Array [
    [Function],
  ],
]
`;

exports[`TasksTable bulk actions handles bulkCancelById requests that are not cancellable 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "message": Object {
          "message": "Not all the selected tasks can be cancelled",
          "type": "warning",
        },
      },
      "type": "TOASTS_ADD",
    },
  ],
]
`;

exports[`TasksTable bulk actions handles bulkCancelBySearch requests 1`] = `
Array [
  Array [
    Object {
      "type": "TASKS_CANCEL_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "message": Object {
          "message": "Canceling selected tasks, this might take a while",
          "type": "info",
        },
      },
      "type": "TOASTS_ADD",
    },
  ],
]
`;

exports[`TasksTable bulk actions handles bulkForceCancelById requests 1`] = `
Array [
  Array [
    Object {
      "type": "TASKS_FORCE_CANCEL_REQUEST",
    },
  ],
  Array [
    Object {
      "type": "TASKS_FORCE_CANCEL_SUCCESS",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "message": Object {
          "message": "2 task(s) cancelled with force",
          "type": "success",
        },
      },
      "type": "TOASTS_ADD",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "message": Object {
          "message": "4 task(s) are already stopped",
          "type": "warning",
        },
      },
      "type": "TOASTS_ADD",
    },
  ],
  Array [
    "TASKS_TABLE",
  ],
  Array [
    [Function],
  ],
]
`;

exports[`TasksTable bulk actions handles bulkForceCancelById requests that are stopped 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "message": Object {
          "message": "1 task(s) are already stopped",
          "type": "warning",
        },
      },
      "type": "TOASTS_ADD",
    },
  ],
]
`;

exports[`TasksTable bulk actions handles bulkForceCancelBySearch requests 1`] = `
Array [
  Array [
    Object {
      "type": "TASKS_FORCE_CANCEL_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "message": Object {
          "message": "Canceling with force selected tasks, this might take a while",
          "type": "info",
        },
      },
      "type": "TOASTS_ADD",
    },
  ],
]
`;

exports[`TasksTable bulk actions handles bulkResumeById requests that are not resumable 1`] = `
Array [
  Array [
    Object {
      "payload": Object {
        "message": Object {
          "message": "Not all the selected tasks can be resumed",
          "type": "warning",
        },
      },
      "type": "TOASTS_ADD",
    },
  ],
]
`;

exports[`TasksTable bulk actions handles bulkResumeById requests that fail 1`] = `
Array [
  Array [
    Object {
      "type": "TASKS_RESUME_REQUEST",
    },
  ],
  Array [
    Object {
      "error": [Error: Network Error],
      "type": "TASKS_RESUME_FAILURE",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "message": Object {
          "message": "Cannot resume tasks at the moment Error: Network Error",
          "type": "error",
        },
      },
      "type": "TOASTS_ADD",
    },
  ],
]
`;

exports[`TasksTable bulk actions handles bulkResumeBySearch requests 1`] = `
Array [
  Array [
    Object {
      "type": "TASKS_RESUME_REQUEST",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "message": Object {
          "message": "Resuming selected tasks, this might take a while",
          "type": "info",
        },
      },
      "type": "TOASTS_ADD",
    },
  ],
]
`;

exports[`TasksTable bulk actions handles resumable bulkResumeById requests 1`] = `
Array [
  Array [
    Object {
      "type": "TASKS_RESUME_REQUEST",
    },
  ],
  Array [
    Object {
      "type": "TASKS_RESUME_SUCCESS",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "message": Object {
          "message": "I am resumed Task execution was resumed",
          "type": "success",
        },
      },
      "type": "TOASTS_ADD",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "message": Object {
          "message": "I am failed Task execution could not be resumed",
          "type": "error",
        },
      },
      "type": "TOASTS_ADD",
    },
  ],
  Array [
    "TASKS_TABLE",
  ],
  Array [
    [Function],
  ],
]
`;

exports[`TasksTable bulk actions handles skipped bulkCancelById requests 1`] = `
Array [
  Array [
    Object {
      "type": "TASKS_CANCEL_REQUEST",
    },
  ],
  Array [
    Object {
      "type": "TASKS_CANCEL_SUCCESS",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "message": Object {
          "message": "I am skipped Task execution task has to be cancellable",
          "type": "warning",
        },
      },
      "type": "TOASTS_ADD",
    },
  ],
]
`;

exports[`TasksTable bulk actions handles skipped bulkResumeById requests 1`] = `
Array [
  Array [
    Object {
      "type": "TASKS_RESUME_REQUEST",
    },
  ],
  Array [
    Object {
      "type": "TASKS_RESUME_SUCCESS",
    },
  ],
  Array [
    Object {
      "payload": Object {
        "message": Object {
          "message": "I am skipped Task execution task has to be resumable",
          "type": "warning",
        },
      },
      "type": "TOASTS_ADD",
    },
  ],
]
`;
